google/cloud/pubsub/internal/tracing_batch_callback.cc

```diff
@@ -0,0 +1,146 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/internal/batch_callback.h"
+#include "google/cloud/pubsub/subscription.h"
+#include "google/cloud/pubsub/version.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/pubsub/internal/message_propagator.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "opentelemetry/context/propagation/text_map_propagator.h"
+#include "opentelemetry/trace/propagation/http_trace_context.h"
+#include "opentelemetry/trace/semantic_conventions.h"
+#include "opentelemetry/trace/span_startoptions.h"
+#include <google/pubsub/v1/pubsub.pb.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace {
+
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> StartSubscribeSpan(
+    google::pubsub::v1::ReceivedMessage const& message,
+    pubsub::Subscription const& subscription,
+    std::shared_ptr<
+        opentelemetry::context::propagation::TextMapPropagator> const&
+        propagator) {
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kConsumer;
+  auto m = pubsub_internal::FromProto(message.message());
+  auto context = ExtractTraceContext(m, *propagator);
+  auto producer_span_context =
+      opentelemetry::trace::GetSpan(context)->GetContext();
+  if (producer_span_context.IsSampled() && producer_span_context.IsValid()) {
+    options.parent = producer_span_context;
+  }
+
+  auto span = internal::MakeSpan(subscription.subscription_id() + " subscribe",
+                                 options);
+  return span;
+}
+
+}  // namespace
+
+/**
+ * Tracing batch callback implementation.
+ * */
+class TracingBatchCallback : public BatchCallback {
```

Comment:
Can this be in the anonymous namespace too?

Reply:
Yes